zip_0143.py

```diff
@@ -0,0 +1,164 @@
+    for x in tx.vout:
+        digest.update(bytes(x))
+    return digest.digest()
+
+def getHashJoinSplits(tx):
+    digest = blake2b(digest_size=32, person=b'ZcashJSplitsHash')
+    for jsdesc in tx.vJoinSplit:
+        digest.update(bytes(jsdesc))
+    digest.update(tx.joinSplitPubKey)
+    return digest.digest()
+
+
+# Currently assumes the nHashType is SIGHASHALL
+# and that there are no joinSplits
+def signature_hash(scriptCode, tx, nIn, nHashType, amount, consensusBranchId):
+    hashPrevouts = b'\x00'*32
+    hashSequence = b'\x00'*32
+    hashOutputs = b'\x00'*32
+    hashJoinSplits = b'\x00'*32
+
+    if not nHashType & SIGHASH_ANYONECANPAY:
+        hashPrevouts = getHashPrevouts(tx)
+
+    if (not nHashType & SIGHASH_ANYONECANPAY) and \
+        (nHashType & 0x1f) != SIGHASH_SINGLE and \
+        (nHashType & 0x1f) != SIGHASH_NONE:
+        hashSequence = getHashSequence(tx)
+
+    if (nHashType & 0x1f) != SIGHASH_SINGLE and \
+        (nHashType & 0x1f) != SIGHASH_NONE:
+        hashOutputs = getHashOutputs(tx)
+    elif (nHashType & 0x1f) == SIGHASH_SINGLE and \
+        0 <= nIn and nIn < len(tx.vout):
+        digest = blake2b(digest_size=32, person=b'ZcashOutputsHash')
+        digest.update(bytes(tx.vout[nIn]))
+        hashOutputs = digest.digest()
+
+    if len(tx.vJoinSplit):
+        hashJoinSplits = getHashJoinSplits(tx)
+
+    digest = blake2b(
+        digest_size=32,
+        person=b'ZcashSigHash' + struct.pack('<I', consensusBranchId),
+    )
+
+    digest.update(struct.pack('<I', tx.header()))
+    digest.update(struct.pack('<I', tx.nVersionGroupId))
+    digest.update(hashPrevouts)
+    digest.update(hashSequence)
+    digest.update(hashOutputs)
+    digest.update(hashJoinSplits)
+    digest.update(struct.pack('<I', tx.nLockTime))
+    digest.update(struct.pack('<I', tx.nExpiryHeight))
+    digest.update(struct.pack('<I', nHashType))
+
+    if nIn != NOT_AN_INPUT:
+        digest.update(bytes(tx.vin[nIn].prevout))
+        digest.update(bytes(scriptCode))
+        digest.update(struct.pack('<Q', amount))
+        digest.update(struct.pack('<I', tx.vin[nIn].nSequence))
+
+    return digest.digest()
+
+
+def main():
+    args = render_args()
+
+    from random import Random
+    rng = Random(0xabad533d)
+    def randbytes(l):
+        ret = []
+        while len(ret) < l:
+            ret.append(rng.randrange(0, 256))
+        return bytes(ret)
+    rand = Rand(randbytes)
+
+    consensusBranchId = 0x5ba81b19 # Overwinter
+
+    test_vectors = []
+    for i in range(10):
+        tx = Transaction(rand, OVERWINTER_TX_VERSION)
+        scriptCode = Script(rand)
+        nIn = rand.u8() % (len(tx.vin) + 1)
+        if nIn == len(tx.vin):
+            nIn = NOT_AN_INPUT
+        nHashType = SIGHASH_ALL if nIn == NOT_AN_INPUT else rand.a([
+            SIGHASH_ALL,
+            SIGHASH_NONE,
+            SIGHASH_SINGLE,
+        ])
+        amount = rand.u64() % MAX_MONEY
```

Comment:
This is very pedantic: ``rand.u64() % (MAX_MONEY+1)``.

(This would mainly act as documentation that the range is 0..MAX_MONEY inclusive.)